src/devices/bus/ti99/internal/992board.cpp

```diff
@@ -458,13 +458,13 @@ void io992_device::device_start()
 
 READ8_MEMBER(io992_device::cruread)
 {
-	int address = offset << 4;
+	int address = offset << 1;
```

Comment:
No, the handlers will see the right-justified hardware address, since the emumem core applies the address shift first. If the software address were provided instead, it would cause unnecessary complications for hooking up generic devices to CPUs like the TMS34010.

Reply:
It is OK for me to use the hardware address (will do a <<1 in the peripheral cards if required), but your changes were not consistent (see tms99xx.cpp, tms9995.cpp where you used the even-only software addresses with cruaddr +=2).

Reply:
The address shift parameter means using the software address within address maps and the CPU core itself, for TMS34010 and other CPUs as well as the TMS99xx. Addresses are shifted before passing them to the peripheral handlers because most of those need only use a few low-order bits anyway. So just like your average 16-bit bus peripheral or 8-bit peripheral mapped to even/odd bytes on a 16-bit bus, a LS259 or something mapped to CRU space at 400 to 40F will only see addresses between 0 and 7.

You'll see a few 16-bit handlers in MAME that run tests on offset << 1 because they're trying to trap some specific accesses to hack around an emulation problem. Handlers ordinarily shouldn't need to shift addresses like that, though.

Reply:
Ah ... that is, the CPU sets e.g. the 0x1100 CRU address, and due to this feature in the address map, the actual value arriving at the read/write handler is 0x0880? Sounds a bit like hidden magic that one must be aware of.

I used the shift <<1 in the handlers to match the handling with the specifications, e.g. when the manual of the disk controller says that 1100 is the card ROM, 1102 is the motor etc. Typically, all those CRU address references are on software base (R12).

Reply:
I added some comments to cpu/tms9900/tms9900.cpp regarding the "hidden magic" of address shifting and other considerations which CRU handlers need to take into account.

Reply:
In general, I agree with your patch, as long as it is properly documented. If you want to merge your changes I'd prefer you do it during the next two weeks, since I have some free days until then.